{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetch Complete genome from the query string which contains only lineage information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import Module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import requests\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fetch Data  Genome "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n"
     ]
    }
   ],
   "source": [
    "c = 0   \n",
    "file1 = open(\"HPV_Genome.fasta\", \"w\") # Open file which contains the complete genome of HPV16 virus which have a lineage information \n",
    "set1 =set()     # empty set for fetching only unique acession ID \n",
    "Entrez.email = 'pradumnpc@gmail.com'\n",
    "\n",
    "# Using NCBI eutils fetch the all id which contains the complete genome for HPV 16 virus \n",
    "\n",
    "x=requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nuccore&term=complete genome[Title] AND \"Human papillomavirus type 16\"[Organism]&retmax=700')\n",
    "m=x.text  # URL request into text format for further processiong \n",
    "reg=re.finditer('<Id>(\\d+)</Id>',m)  # regular expression which fetch only accesion id from given html page in list format \n",
    "i=0\n",
    " # declare a list which contains the different lineage annotation for same lineage after we itereate all this list to find out exact match for the lineage\n",
    "A1 = ['genotype: A1','lineage A sublineage A1 (European, E)','lineage A1'] \n",
    "A2 = ['lineage A sublineage A2 (European, E)','lineage A2','genotype: A2']  \n",
    "D3 = ['genotype: D3','lineage D3']\n",
    "\n",
    "for i in reg:  # itereate loop on all accesion id \n",
    "    id=i.group(1)  # get the only accesion ID \n",
    "    print(c)\n",
    "    c+=1\n",
    "    # url wwhich fetch the genebank file of perticular accesion ID \n",
    "    \n",
    "    url=requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id={}&rettype=gb'.format(id))\n",
    "    content=url.text  # convert the url request into text format \n",
    "    lineage = re.search(r'/note=\"(.*)\"',content)  # regular expression for fetching lineage info of perticular genebank ID \n",
    "    if lineage:   #  condition true only when get the lineage information is fetch \n",
    "        set1.add(lineage.group(1))\n",
    "        cond = lineage.group(1) \n",
    "#         print(cond)\n",
    "        for j in A1:  # loop over the A1 list which contains the different lineage annotation for A1 lineage \n",
    "            if j == cond:  \n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text') # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')  \n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq) \n",
    "                file1.write('>A1_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "\n",
    "        for j in A2:\n",
    "            if j == cond:\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>A2_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage A3':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>A3_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage A4':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>A4_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage B1':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>B1_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage B2':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>B2_{}_l1\\n'.format(rec.id)+sequence +'\\n') \n",
    "        if cond == 'lineage B3':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>B3_{}_l1\\n'.format(rec.id)+sequence +'\\n')  \n",
    "        if cond == 'lineage B4':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>B4_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage C1':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>C1_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage C3':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>C3_{}_l1\\n'.format(rec.id)+sequence +'\\n')   \n",
    "        if cond == 'lineage C4':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>C4_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage D1':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>D1_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage D2':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>D2_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        for j in D3:\n",
    "            if cond == j:\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>D3_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "        if cond == 'lineage D4':\n",
    "                gb_acc1 = Entrez.efetch(db='nucleotide', id=id, rettype='gb', retmode='text')   # using biopython fetch genebank which contains the linaege \n",
    "                rec = SeqIO.read(gb_acc1, 'genbank')\n",
    "                print(rec.id)\n",
    "                sequence = str(rec.seq)\n",
    "                file1.write('>D4_{}_l1\\n'.format(rec.id)+sequence +'\\n')\n",
    "\n",
    "file1.close()            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Asian American variant',\n",
       " 'Asian variant',\n",
       " 'E-C109/G350',\n",
       " 'E6 ORF from 65 to 559; putative',\n",
       " 'European variant',\n",
       " 'HpV16gp1',\n",
       " 'genotype: A1',\n",
       " 'genotype: A2',\n",
       " 'genotype: D3',\n",
       " 'genotype: HPV16',\n",
       " 'lineage A',\n",
       " 'lineage A sublineage A1 (European, E)',\n",
       " 'lineage A sublineage A2 (European, E)',\n",
       " 'lineage A1',\n",
       " 'lineage A2',\n",
       " 'lineage A3',\n",
       " 'lineage A4',\n",
       " 'lineage Af-1',\n",
       " 'lineage Af-2',\n",
       " 'lineage Asian-American (AA)',\n",
       " 'lineage Asian-American (AA1)',\n",
       " 'lineage Asian-American (AA2)',\n",
       " 'lineage Asian-American (NA1)',\n",
       " 'lineage B1',\n",
       " 'lineage B2',\n",
       " 'lineage B3',\n",
       " 'lineage B4',\n",
       " 'lineage C',\n",
       " 'lineage C1',\n",
       " 'lineage C3',\n",
       " 'lineage C4',\n",
       " 'lineage D1',\n",
       " 'lineage D2',\n",
       " 'lineage D3',\n",
       " 'lineage D4',\n",
       " 'lineage E',\n",
       " 'lineage E (Asian)',\n",
       " 'lineage E (prototype 1)',\n",
       " 'lineage E (prototype 2)',\n",
       " 'lineage E(As)',\n",
       " 'lineage E(prototype)',\n",
       " 'lineage NA',\n",
       " 'oncogene',\n",
       " 'putative transforming protein',\n",
       " 'replaced G with A',\n",
       " 'start codon is not identified',\n",
       " 'start codon not determined',\n",
       " 'start codon not identified',\n",
       " 'subtype: Asian American',\n",
       " 'type: 16',\n",
       " 'type: East Asian'}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print unique lineage annotation for HPV16 virus \n",
    "print(set1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# recheck for unique entry in dataset \n",
    "f = open(\"HPV_Genome_no_Duplicate1.fasta\", \"r\")\n",
    "file = f.readlines()\n",
    "sequences = []\n",
    "g = []\n",
    "seq = \"\"\n",
    "for f in file:\n",
    "    if f.startswith('>'):\n",
    "        g.append(f)\n",
    "    if not f.startswith('>'):\n",
    "        f = f.replace(\" \", \"\")\n",
    "        f = f.replace(\"\\n\", \"\")\n",
    "        seq = seq + f\n",
    "    else:\n",
    "        sequences.append(seq)\n",
    "        seq = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Check the frequency of clade in whole genome dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A1= 104 A2= 23 A3= 7 A4= 62 B1= 5 B2= 4 B3= 3 B4= 4 C1= 8 C2= 1 8 C3= 3 C4= 4 D1= 7 D2= 7 D3= 17 D4= 4\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "s = []\n",
    "A1 = 0\n",
    "A2 = 0\n",
    "A3 = 0\n",
    "A4 = 0\n",
    "B1 = 0 \n",
    "B2 = 0 \n",
    "B3 = 0\n",
    "B4 = 0\n",
    "C1 = 0 \n",
    "C2 = 0\n",
    "C3 = 0\n",
    "C2 = 0\n",
    "C4 = 0\n",
    "D1 = 0\n",
    "D2 = 0 \n",
    "D3 = 0 \n",
    "D4 = 0\n",
    "for i in g:\n",
    "    id1 = re.search(r'>.*_(\\w\\d)',i)\n",
    "    id1 = id1.group(1)\n",
    "    if id1 == 'A1':\n",
    "        A1+=1\n",
    "    if id1 == 'A2':\n",
    "        A2+=1\n",
    "    if id1 == 'A3':\n",
    "        A3+=1\n",
    "    if id1 == 'A4':\n",
    "        A4+=1\n",
    "    if id1 == 'B1':\n",
    "        B1+=1\n",
    "    if id1 == 'B2':\n",
    "        B2+=1\n",
    "    if id1 == 'B3':\n",
    "        B3+=1\n",
    "    if id1 == 'B4':\n",
    "        B4+=1\n",
    "    if id1 == 'C1':\n",
    "        C1+= 1\n",
    "    if id1 == 'C2':\n",
    "        C2+=1\n",
    "    if id1 == 'C3':\n",
    "        C3+=1\n",
    "    if id1 == 'C4':\n",
    "        C4+=1\n",
    "    if id1 == 'D1':\n",
    "        D1+=1\n",
    "    if id1 == 'D2':\n",
    "        D2+=1\n",
    "    if id1 == 'D3':\n",
    "        D3+=1\n",
    "    if id1 == 'D4':\n",
    "        D4+=1\n",
    "\n",
    "        \n",
    "print('A1=',A1,'A2=',A2,'A3=',A3,'A4=',A4,'B1=',B1,'B2=',B2,'B3=',B3,'B4=',B4,'C1=',C1,'C2=',C2,C1,'C3=',C3,'C4=',C4,'D1=',D1,'D2=',D2,'D3=',D3,'D4=',D4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a label list for itol phylogentic tree visualisation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "file = open('HPV_Genome_no_Duplicate1.fasta').readlines()\n",
    "id1 = open('HPV_Genome_ID_No_duplicate1.txt','w')\n",
    "pat =('>\\w\\d_.*')\n",
    "c = 0\n",
    "[id1.write(i.rstrip()+\" label #ee38ff bold 2\"+\"\\n\") for i in file if re.match('>.*_A1',i)]\n",
    "[id1.write(i.rstrip()+\" label #ee38ff bold 2\"+\"\\n\") for i in file if re.match('>.*_A2',i)]\n",
    "[id1.write(i.rstrip()+\" label #ee38ff bold 2\"+\"\\n\") for i in file if re.match('>.*_A3',i)]\n",
    "[id1.write(i.rstrip()+\" label #ee38ff bold 2\"+\"\\n\") for i in file if re.match('>.*_A4',i)]\n",
    "[id1.write(i.rstrip()+\" label #2afa38 bold 2\"+\"\\n\") for i in file if re.match('>.*_B1',i)]\n",
    "[id1.write(i.rstrip()+\" label #2afa38 bold 2\"+\"\\n\") for i in file if re.match('>.*_B2',i)]\n",
    "[id1.write(i.rstrip()+\" label #2afa38 bold 2\"+\"\\n\") for i in file if re.match('>.*_B3',i)]\n",
    "[id1.write(i.rstrip()+\" label #2afa38 bold 2\"+\"\\n\") for i in file if re.match('>.*_B4',i)]\n",
    "[id1.write(i.rstrip()+\" label #04FC97 bold 2\"+\"\\n\") for i in file if re.match('>.*_C1',i)]\n",
    "[id1.write(i.rstrip()+\" label #04FC97 bold 2\"+\"\\n\") for i in file if re.match('>.*_C2',i)]\n",
    "[id1.write(i.rstrip()+\" label #04FC97 bold 2\"+\"\\n\") for i in file if re.match('>.*_C3',i)]\n",
    "[id1.write(i.rstrip()+\" label #04FC97 bold 2\"+\"\\n\") for i in file if re.match('>.*_C4',i)]\n",
    "[id1.write(i.rstrip()+\" label #FD0377 bold 2\"+\"\\n\") for i in file if re.match('>.*_D1',i)]\n",
    "[id1.write(i.rstrip()+\" label #FD0377 bold 2\"+\"\\n\") for i in file if re.match('>.*_D2',i)]\n",
    "[id1.write(i.rstrip()+\" label #FD0377 bold 2\"+\"\\n\") for i in file if re.match('>.*_D3',i)]\n",
    "[id1.write(i.rstrip()+\" label #FD0377 bold 2\"+\"\\n\") for i in file if re.match('>.*_D4',i)]\n",
    "\n",
    "\n",
    "id1.close()\n",
    "print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print Itol label information for dataset \n",
    "file1 = open('HPV_Genome_ID_List.txt')\n",
    "for i in file1:\n",
    "    id2 = re.search('>(.*)',i)\n",
    "    print(id2.group(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
